<a href="https://www.kaggle.com/code/ahmedfakhar123/pandas-project-ai-dataset-preprocessing-eda?scriptVersionId=334838898" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

---
# 🐼 Pandas Project — AI Dataset Preprocessing & Exploratory Data Analysis

>Learn how to preprocess and explore real-world data using **Pandas**. This project covers data cleaning, handling missing values, feature engineering, exploratory data analysis (EDA), and visualization using the Titanic dataset to prepare data for future machine learning tasks.



---
# 📌 1. Introduction

Data preprocessing and ***Exploratory Data Analysis (EDA)*** are fundamental steps in every data science and machine learning workflow. Before building any model, it is essential to understand the dataset, identify data quality issues, clean inconsistencies, and explore meaningful patterns.

In this project, we will use the ***Titanic*** dataset to perform a complete preprocessing and EDA workflow using ***Pandas***. The notebook covers data exploration, handling missing values, removing duplicate records, feature engineering, statistical analysis, and data visualization.

By the end of this project, the dataset will be ***clean, well-understood, and ready for machine learning***, while also revealing valuable insights into the factors that influenced passenger survival.

<div align="center">
    <img src="https://media.geeksforgeeks.org/wp-content/uploads/20260401165333112780/exploratory_data_analysis_eda_2.webp" width=500>
    <img src="https://media.geeksforgeeks.org/wp-content/uploads/20260203154034613956/data_preprocessing.webp" width=500>
</div>

## 🎯 Objectives
1. ***📥 Load and inspect the dataset***
2. ***🔍 Explore its structure and statistics***
3. ***🧹 Clean and preprocess the data***
4. ***⚙️ Perform feature engineering***
5. ***📊 Conduct Exploratory Data Analysis (EDA)***
6. ***📈 Visualize trends and relationships***
7. ***💡 Extract meaningful insights***
8. ***💾 Export the cleaned dataset***

---

# 📂 2. Import Libraries

To perform ***data preprocessing***, ***analysis***, and ***visualization***, we first import the required Python libraries. Each library provides ***specialized tools*** that simplify working with datasets, from loading and cleaning data to generating ***informative visualizations.***

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

---

# 📥 3. Load the Dataset

The first step is to load the ***Titanic dataset*** into a `Pandas DataFrame`. Once loaded, we can begin ***exploring*** its structure, examine the available features, and prepare it for ***preprocessing*** and ***exploratory data analysis.***

In [2]:
df = pd.read_csv("/kaggle/input/datasets/yasserh/titanic-dataset/Titanic-Dataset.csv")
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


---

# 👀 4. Dataset Overview
Before cleaning or analyzing the data, it's important to understand the ***dataset's structure*** and ***contents.*** This section provides a ***quick overview*** of the dataset, helping us identify its ***size***, ***features***, ***data types***, and ***basic statistical characteristics***.

## 4.1 Display First & Last Rows
View the ***first*** and ***last*** few records of the dataset to get a ***quick understanding*** of its structure, values, and overall format.

In [3]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [4]:
df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


## 4.2 Dataset Shape
Determine the ***number of rows and columns*** to understand the ***size and dimensions*** of the dataset.

In [5]:
df.shape

(891, 12)

This shows that `df` has ***891 rows*** and ***12 columns***

## 4.3 Column Names

Display all ***column names*** to identify the ***available features*** and understand what information the dataset contains.

In [6]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

## 4.4 Data Types

Examine the ***data type*** of each column to distinguish between ***numerical, categorical, and text-based*** features, ensuring they are suitable for analysis.

In [7]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

## 4.5 Dataset Information

Get a ***comprehensive summary*** of the dataset, including the total ***number of entries***, ***non-null*** values, ***memory usage***, and ***data types*** for each column.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


## 4.6 Statistical Summary

Generate ***descriptive statistics*** for numerical features, including ***count***, ***mean***, ***standard deviation***, ***minimum***, ***maximum***, and ***quartile values,*** to better understand the data distribution.

In [9]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


---

# 🔎 5. Initial Data Exploration
Before cleaning the dataset, it's important to perform an ***initial exploration*** to assess its overall quality and identify potential issues. This step helps us understand the ***completeness***, ***uniqueness***, and ***structure of the data***, allowing us to make informed preprocessing decisions.

## 5.1 Check Missing Values
Identify columns that contain missing (`NaN`) values and determine how much data is missing in each. This helps decide whether to ***fill***, ***remove***, or ***retain missing entries*** during preprocessing.


In [10]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

This shows that `Age` column has ***177***, `Cabin` column has ***687***, and `Embarked` column has ***2*** `NaN` values.

## 5.2 Check Duplicate Records
Detect duplicate rows that may introduce ***bias*** or ***redundancy*** into the analysis. Removing unnecessary duplicates helps improve ***data quality*** and ensures more ***reliable results***.

In [11]:
df.duplicated().sum()

np.int64(0)

Here `0` shows that no ***row*** is repeated.

## 5.3 Unique Values per Column
Examine the number of unique values in ***each column*** to better understand the ***dataset's diversity***, ***identify categorical features***, and spot any unusual or ***inconsistent values***.

In [12]:
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

## 5.4 Memory Usage
Measure the amount of memory consumed by the dataset. This is particularly ***useful*** when working with ***large datasets***, as it helps ***optimize memory usage*** and ***improve processing efficiency***.

In [13]:
df.memory_usage()

Index           132
PassengerId    7128
Survived       7128
Pclass         7128
Name           7128
Sex            7128
Age            7128
SibSp          7128
Parch          7128
Ticket         7128
Fare           7128
Cabin          7128
Embarked       7128
dtype: int64

This is seperate memory taken by ***each column***. However, the ***total memory*** taken by `df` is:

In [14]:
df.memory_usage().sum()

np.int64(85668)


---

## ⚙️ 7. Feature Engineering

* Create New Features *(e.g., Family Size = SibSp + Parch + 1)*
* Categorize Age Groups
* Extract Passenger Titles from Names *(optional)*

---

## 📊 8. Exploratory Data Analysis (EDA)

### 🎯 Survival Analysis

* Overall Survival Rate
* Survival by Gender
* Survival by Passenger Class
* Survival by Age Group
* Survival by Embarkation Port

### 👥 Passenger Demographics

* Age Distribution
* Gender Distribution
* Passenger Class Distribution
* Family Size Distribution

### 💰 Fare Analysis

* Fare Distribution
* Fare by Passenger Class
* Fare vs Survival

### 🔗 Relationship Analysis

* Correlation Matrix
* Pairwise Relationships *(optional)*

---

## 📈 9. Data Visualization

* 📊 Bar Charts
* 📈 Histograms
* 📦 Box Plots
* ⚪ Scatter Plots
* 🔥 Heatmap
* 🥧 Pie Charts *(optional)*

---

## 💡 10. Key Insights

Example:

* Female passengers had a significantly higher survival rate.
* First-class passengers survived more often.
* Younger passengers generally had better survival chances.
* Higher ticket fares were associated with increased survival.

---

## 💾 11. Export the Cleaned Dataset

---

## ✅ 12. Conclusion

---

## 📚 13. Next Steps

* Build a Machine Learning model
* Perform Feature Selection
* Train Classification Algorithms

---

## ❤️ Support

If you found this notebook helpful, consider **upvoting ⭐** it and sharing your feedback. Contributions and suggestions are always welcome!

---
---